aea/protocols/generator/validate.py

```diff
@@ -18,60 +18,183 @@
+    content_type = content_type.strip()
+
+    if not content_type.startswith("pt:set"):
+        return False
+
+    if not _has_brackets(content_type):
+        return False
+
+    sub_types = _get_sub_types_of_compositional_types(content_type)
+    if len(sub_types) != 1:
+        return False
+
+    sub_type = sub_types[0]
+    return _is_valid_pt(sub_type)
+
+
+def _is_valid_list(content_type: str) -> bool:
+    content_type = content_type.strip()
+
+    if not content_type.startswith("pt:list"):
+        return False
+
+    if not _has_brackets(content_type):
+        return False
+
+    sub_types = _get_sub_types_of_compositional_types(content_type)
+    if len(sub_types) != 1:
+        return False
+
+    sub_type = sub_types[0]
+    return _is_valid_pt(sub_type)
+
+
+def _is_valid_dict(content_type: str) -> bool:
+    content_type = content_type.strip()
+
+    if not content_type.startswith("pt:dict"):
+        return False
+
+    if not _has_brackets(content_type):
+        return False
+
+    sub_types = _get_sub_types_of_compositional_types(content_type)
+    if len(sub_types) != 2:
+        return False
+
+    sub_type_1 = sub_types[0]
+    sub_type_2 = sub_types[1]
+    return _is_valid_pt(sub_type_1) and _is_valid_pt(sub_type_2)
+
+
+def _is_valid_union(content_type: str) -> bool:
+    content_type = content_type.strip()
+
+    if not content_type.startswith("pt:union"):
+        return False
+
+    if not _has_brackets(content_type):
+        return False
+
+    sub_types = _get_sub_types_of_compositional_types(content_type)
+    for sub_type in sub_types:
+        if not (
+            _is_valid_ct(sub_type)
+            or _is_valid_pt(sub_type)
+            or _is_valid_set(sub_type)
+            or _is_valid_list(sub_type)
+            or _is_valid_dict(sub_type)
+        ):
+            return False
+
+    return True
+
+
+def _is_valid_optional(content_type: str) -> bool:
+    content_type = content_type.strip()
+
+    if not content_type.startswith("pt:optional"):
+        return False
+
+    if not _has_brackets(content_type):
+        return False
 
+    sub_types = _get_sub_types_of_compositional_types(content_type)
+    if len(sub_types) != 1:
+        return False
 
-# ToDo other validation functions
+    sub_type = sub_types[0]
+    return (
+        _is_valid_ct(sub_type)
+        or _is_valid_pt(sub_type)
+        or _is_valid_set(sub_type)
+        or _is_valid_list(sub_type)
+        or _is_valid_dict(sub_type)
+        or _is_valid_union(sub_type)
+    )
+
+
+def _is_valid_content_type_format(content_type: str) -> bool:
+    return (
+        _is_valid_ct(content_type)
+        or _is_valid_pt(content_type)
+        or _is_valid_set(content_type)
+        or _is_valid_list(content_type)
+        or _is_valid_dict(content_type)
+        or _is_valid_union(content_type)
+        or _is_valid_optional(content_type)
+    )
 
 
 def validate(protocol_specification: ProtocolSpecification) -> Tuple[bool, str]:
```

Comment:
This function could be further broken down (essentially along each comment block)

Reply:
Done